### PR TITLE
Add CIP-66 (CELO denominated fees tx type)

### PR DESCRIPTION
**Pull request by karlb**

### CODE REVIEW

File: CIPs/cip-0066.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+---
+cip: 66
+title: "New Transaction Type: Celo Denominated Fees"
+author: Karl Bartel (@karlb)
+discussions-to:
```
Comment (piersy) [inline CIPs/cip-0066.md:5]: Should this be set?

Comment (karlb) [reply CIPs/cip-0066.md:5]: Eventually yes, but we need to make a forum post first to have the link and I wanted to get this reviewed internally before we post to the forum.

File: CIPs/cip-0066.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+The `signatureYParity, signatureR, signatureS` elements of this transaction represent a secp256k1 signature over `keccak256(0x7a || rlp([chainId, nonce, maxPriorityFeePerGas, maxFeePerGas, gasLimit, to, value, data, accessList, feeCurrency, maxFeeInFeeCurrency]))`.
+
+The [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) `ReceiptPayload` for this transaction is `rlp([status, cumulativeGasUsed, logsBloom, logs, feeInFeeCurrency])`.
+
+The new `feeInFeeCurrency` receipt field is also included in RPC responses for CIP-66 txs.
+
+## Rationale
+
+### Rates vs. absolute values
+The fields `maxFeeInFeeCurrency` and `feePerGasInFeeCurrency` are meant to convey information about conversion rates (allowed conversion rates for the former and used rate for the latter). Actually encoding this information as conversion rate fields would not be straightforward, as only integer numbers are supported natively by the EVM. There are many ways to encode such information, but none is as accurate and easy to use as plain integer values for the resulting totals.
```
Comment (aaronmgdr) [inline CIPs/cip-0066.md:73]: `feePerGasInFeeCurrency` is not mentioned anywhere else but here

File: CIPs/cip-0066.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+### Including `feeInFeeCurrency` in the receipt
+It must be obvious exactly how much has been paid in fees for a tx and since the fees are paid in `feeCurrency`, the total fees for a tx must also be available denominated in that fee currency. Without modifying the receipt, the `feeInFeeCurrency` could be derived from the tx, the base fee and the conversion rate at the beginning of the block.
+
+But the conversion rate is stored in the block state, which gets pruned from non-archive nodes after a while. This would mean that `feeInFeeCurrency` is returned via JSON RPC after the tx was executed, but then "disappears" from the JSON receipt after a while. This is both surprising (receipt responses should not change over time) and problematic, since it makes the exact fee amount unavailable. To avoid this, we chose to include the `feeInFeeCurrency` in the RLP-encoded receipt.
```
Comment (aaronmgdr) [inline CIPs/cip-0066.md:78]: including this in reciept is nice!

File: CIPs/cip-0066.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+While this introduces a difference to EIP-1559, the downsides are limited because
+* Most tools access the receipts via JSON RPC, where the additional field won't do any harm.
+* The field is hidden inside the [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) compliant receipt payload.
+
+### `feeInFeeCurrency` vs `effectiveGasPriceInFeeCurrency`
+
+Considering that JSON RPC responses contain an `effectiveGasPrice` field, it would be natural to add an `effectiveGasPriceInFeeCurrency` field instead of the suggested `feeInFeeCurrency` field. This does not work due to the limited accuracy of defining prices per gas in fee currency. For tokens with a low number of decimals, this could not accurately represent all relevant prices, thereby either limiting the allowed prices or misrepresenting the actually paid price.
```
Comment (aaronmgdr) [inline CIPs/cip-0066.md:86]: im confused as to what should be done

File: CIPs/cip-0066.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+## Specification
+
+### Definitions
+
+- `TransactionType` `122` ⇒ `0x7a`. See [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
+- `||` is the byte/byte-array concatenation operator
+- `//` is integer division (discarding the remainder)
+- `feeCurrency` is the address of the whitelisted currency to be used to pay for gas. In contrast to older tx types, this field can not be omitted to mean "native CELO token".
+- The exchange rate from CELO to `feeCurrency` is specified as a rational number consisting of two `uint256`: `rateNum` / `rateDen`
+- `maxFeeInFeeCurrency` is chosen by the tx sender to limit the exchange rate risk. Txs must only be included if `maxFeeInFeeCurrency >= gasLimit * maxFeePerGas * rateNum // rateDen`
```
Comment (aaronmgdr) [inline CIPs/cip-0066.md:57]: so to send cip66 

a dev will need to set

`maxPriorityFeePerGas`, and `maxFeePerGas`, the same way as as any 1559 tx
`feeCurrency` same as cip64 tx

and then also `maxFeeInFeeCurrency`  -- are you expecting devs/ libraries to be fetching the exchange rates and calculating this number them selves?

File: CIPs/cip-0066.md (hunk 4 above)
Comment (aaronmgdr) [reply CIPs/cip-0066.md:86]: ok i think im understanding better now. thank you

File: CIPs/cip-0066.md (hunk 5 above)
Comment (karlb) [reply CIPs/cip-0066.md:57]: Unfortunately yes, and this is a certainly a downside of this proposal. If you want to set `maxFeeInFeeCurrency` sensibly (you could set it to maxInt and take all exchange rate risk), you'll have to do the equivalent of
```bash
cast call $(
	cast call 0x000000000000000000000000000000000000ce10 "getAddressForStringOrDie(string calldata identifier) returns (address)" SortedOracles
) 'function medianRate(address token) returns (uint256, uint256)' $TOKEN_ADDRESS
```
or get a reasonable exchange rate from a different source.

File: CIPs/cip-0066.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
+---
```
Comment (mcortesi) [inline CIPs/cip-0066.md:1]: We're missing the transaction application rules (pseducode) here.

How gas is payed, how much, what happens to spread, etc. I believe some form of state transition pseudo code would be important

Comment (karlb) [reply CIPs/cip-0066.md:1]: Good point. I added some tx handling pseudocode.

**Pull request merged**